oci/dd-metrics-function.py

```diff
@@ -0,0 +1,150 @@
+import io
+import json
+import logging
+import os
+import gzip
+
+from fdk import context, response
+import requests
+from requests.adapters import HTTPAdapter
+from requests.exceptions import HTTPError
+
+
+logger = logging.getLogger(__name__)
+
+
+OUTPUT_MESSAGE_VERSION = "v1.0"
+
+_max_pool = int(os.environ.get("DD_MAX_POOL", 10))
+_session = requests.Session()
+_session.mount("https://", HTTPAdapter(pool_connections=_max_pool))
+
+
+def _get_serialized_metric_data(raw_metrics: io.BytesIO) -> str:
+    return raw_metrics.getvalue().decode("utf-8")
+
+
+def _generate_metrics_msg(
+    ctx: context.InvokeContext,
+    serialized_metric_data: str,
+) -> str:
+    tenancy_ocid = os.environ.get("TENANCY_OCID")
+
+    if not tenancy_ocid:
+        raise ValueError("Missing environment variable: TENANCY_OCID")
+
+    # Bump OUTPUT_MESSAGE_VERSION any time this
+    # structure gets updated
+    message_dict = {
+        "version": OUTPUT_MESSAGE_VERSION,
+        "payload": {
+            "headers": {
+                "tenancy_ocid": tenancy_ocid,
+                "source_fn_app_ocid": ctx.AppID(),
+                "source_fn_app_name": ctx.AppName(),
+                "source_fn_ocid": ctx.FnID(),
+                "source_fn_name": ctx.FnName(),
+                "source_fn_call_id": ctx.CallID(),
+            },
+            "body": serialized_metric_data,
+        },
+    }
+
+    return json.dumps(message_dict)
+
+
+def _should_compress_payload() -> bool:
+    return os.environ.get("DD_COMPRESS", "false").lower() == "true"
+
+
+def _send_metrics_msg_to_datadog(metrics_message: str) -> str:
+    endpoint = os.environ.get("DD_INTAKE_HOST")
+    api_key = os.environ.get("DD_API_KEY")
+
+    if not endpoint or not api_key:
+        raise ValueError(
+            "Missing one of the following environment variables: DD_INTAKE_HOST, DD_API_KEY"
+        )
+
+    url = f"https://{endpoint}/api/v2/ocimetrics"
+    api_headers = {"content-type": "application/json", "dd-api-key": api_key}
+
+    if _should_compress_payload():
+        serialized = gzip.compress(metrics_message.encode())
+        api_headers["content-encoding"] = "gzip"
+    else:
+        serialized = metrics_message
+
+    http_response = _session.post(url, data=serialized, headers=api_headers)
+    http_response.raise_for_status()
+
+    logger.info(
+        f"Sent payload size={len(metrics_message)} encoding={api_headers.get('content-encoding', None)}"
+    )
+    return http_response.text
+
+
+def handler(ctx: context.InvokeContext, data: io.BytesIO = None):
```

Comment:
nitpick ⛏️ Would be good to have type hint for response of this function even though you have it in the doc string

Reply:
still missing return type 😬